src/openstack_manager.py

```diff
@@ -113,3 +85,196 @@ def list_projects(cloud_config: dict) -> list[Project]:
+        cloud_config: The cloud configuration to connect OpenStack with.
+        github_client: The Github client to interact with Github API.
+        path: Github organisation or repository path.
+        proxies: HTTP proxy settings.
+
+    Raises:
+        ImageBuildError: If there were errors building/creating the image.
+
+    Returns:
+        The OpenStack image object.
+    """
+    try:
+        runner_application = github_client.get_runner_application(path=path, arch=arch)
+    except RunnerBinaryError as exc:
+        raise ImageBuildError("Failed to fetch runner application.") from exc
+
+    try:
+        execute_command(_build_image_command(runner_application, proxies), check_exit=True)
+    except SubprocessError as exc:
+        raise ImageBuildError("Failed to build image.") from exc
+
+    try:
+        conn = _create_connection(cloud_config)
+        arch = "amd64" if runner_application["architecture"] == "x64" else "arm64"
+        existing_image: openstack.image.v2.image.Image
+        for existing_image in conn.search_images(name_or_id=IMAGE_NAME):
+            # images with same name (different ID) can be created and will error during server
+            # instantiation.
+            if not conn.delete_image(name_or_id=existing_image.id, wait=True):
+                raise ImageBuildError("Failed to delete duplicate image on Openstack.")
+        return conn.create_image(
+            name=IMAGE_NAME, filename=IMAGE_PATH_TMPL.format(architecture=arch), wait=True
+        )
+    except OpenStackCloudException as exc:
+        raise ImageBuildError("Failed to upload image.") from exc
+
+
+def create_instance_config(
+    unit_name: str,
+    openstack_image: openstack.image.v2.image.Image,
+    path: GithubPath,
+    github_client: GithubClient,
+) -> InstanceConfig:
+    """Create an instance config from charm data.
+
+    Args:
+        unit_name: The charm unit name.
+        image: Ubuntu image flavor.
+        path: Github organisation or repository path.
+        github_client: The Github client to interact with Github API.
+    """
+    app_name, unit_num = unit_name.rsplit("/", 1)
+    suffix = secrets.token_hex(12)
+    registration_token = github_client.get_runner_registration_token(path=path)
+    return InstanceConfig(
+        name=f"{app_name}-{unit_num}-{suffix}",
+        labels=(app_name, "jammy"),
+        registration_token=registration_token,
+        github_path=path,
+        openstack_image=openstack_image,
+    )
+
+
+class InstanceLaunchError(Exception):
+    """Exception representing an error during instance launch process."""
+
+
+@retry(tries=5, delay=5, max_delay=60, backoff=2, local_logger=logger)
+def create_instance(
+    cloud_config: dict[str, dict],
+    instance_config: InstanceConfig,
+    proxies: Optional[ProxyConfig] = None,
+    dockerhub_mirror: Optional[str] = None,
+    ssh_debug_connections: list[SSHDebugConnection] | None = None,
+) -> openstack.compute.v2.server.Server:
```

Comment:
Perhaps change this return type too

Reply:
Fixed the return types to a more generic one, thank you!